### PR TITLE
nixos/swap: add options option

### LINKED ISSUE #6603: Adding mount options for swap
Is it possible to implement mount options for swap partition? This feature could be useful for machines with SSD only drive(s). User could change to `defaults,discard` options for automatic trimming. [Man pages](http://man7.org/linux/man-pages/man8/swapon.8.html) of this feature.

### PR DESCRIPTION
**Pull request by Artturin**
###### Motivation for this change
Closes #6603

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Artturin** [comment]
tested with
```
disabledModules = [ "config/swap.nix" "tasks/filesystems.nix" ];

imports = let
  pkgsArtturin = builtins.fetchTarball {
    url = "https://github.com/Artturin/nixpkgs/archive/90da50e9969.tar.gz";
  };
in [
  (import "${pkgsArtturin}/nixos/modules/config/swap.nix")
  (import "${pkgsArtturin}/nixos/modules/tasks/filesystems.nix")
];

  swapDevices = [
    { "device" = "/swapfile"; "options" = [ "nofail" "noatime" ]; "priority" =  0; "discardPolicy" = "once"; }
  ];

```
generated entry
`/swapfile none swap nofail,noatime,pri=0,discard=once`

**Comment by cyounkins** [comment]
Thanks for this. I'd just drop the commented line.

**Comment by Artturin** [comment]
oops

**Comment by Artturin** [comment]
please review.

**Comment by ThibautMarty** [comment]
LGTM.

Aside note: you could avoid copy/pasting the `nonEmptyStr` function by reusing the `fileSystems.<name>.options` type.
This is doable with:
```
type = (options'.fileSystems.type.getSubOptions options.loc).options.type
```
where `options'` is the top-level `option` attribute (as argument of the main function in the file, not the `swapCfg` function).

However, this goes a bit in the module system's internal and I don't find any instance of this in nixpkgs (there's one in `nixos/modules/services/networking/searx.nix` but without submodule) and I'm not sure about the `options.loc` argument (changing values does not seem to change the error messages though).
We should find an elegant solution to reuse options and/or types as this is not the first time I see this problem in nixpkgs (and home-manager). This is beyond the scope of this PR anyway.

**Review by ThibautMarty — APPROVED** [review]

**Review by fufexan — APPROVED** [review]

File: nixos/modules/config/swap.nix [hunk 1]
```diff
@@ -5,6 +5,12 @@ with lib;
 
 let
 
+  # From modules/tasks/filesystems.nix
+  addCheckDesc = desc: elemType: check: types.addCheck elemType check
+  // { description = "${elemType.description} (with check: ${desc})"; };
+  isNonEmpty = s: (builtins.match "[ \t\n]*" s) == null;
+  nonEmptyStr = addCheckDesc "non-empty" types.str isNonEmpty;
```
Comment (roberth) [inline nixos/modules/config/swap.nix:12]: Could you add `nonEmptyStr` to `lib/types.nix`? Please keep the `description` simple, like `"non-empty string"`. There's no need for it to be verbose and mechanical like that. Doing so also removes the need for `addCheckDesc`.

**Comment by Artturin** [comment]
the type is not working currently im testing with
```
  disabledModules = [
    "config/swap.nix" 
  ];

  imports =
    let
      pkgsArtturin = builtins.fetchTarball {
        url = "https://github.com/Artturin/nixpkgs/archive/1a57503edc8cb9b7132d07c1f098cd2f13919ab9.tar.gz";
      };
    in
    [
      (import "${pkgsArtturin}/nixos/modules/config/swap.nix")
    ];

  swapDevices = [
    { "device" = "/swapfile"; "options" = [ "nofail" "noatime" ]; "priority" =  0; "discardPolicy" = "once"; }
  ];
```
no build errors but the generated entry is `/swapfile none swap defaults,pri=0,discard=once`

File: lib/types.nix [hunk 2]
```diff
@@ -287,6 +287,12 @@ rec {
       merge = mergeEqualOption;
     };
 
+    nonEmptyStr = elemType:
```
Comment (roberth) [inline lib/types.nix:290]: You don't need `elemType`. `nonEmptyStr` is already a type and doesn't take a type argument.

File: lib/types.nix [hunk 3]
```diff
@@ -287,6 +287,12 @@ rec {
...
+      let string = addCheck (x: str.check x && builtins.match "[ \t\n]*" x == null);
+      in string // {
+        description = "non-empty string " + string.description;
```
Comment (roberth) [inline lib/types.nix:293]:
```suggestion
        description = "non-empty string";
```

**Comment by Artturin** [comment]
this works but i had to apply both of the commits to my local checkout of nixpkgs and rebuild from that 
[`a3c5f0c` (#128032)](https://github.com/NixOS/nixpkgs/pull/128032/commits/a3c5f0cba8fa9c4d9782ef83757be6e4028f54b7)

**Review by roberth — APPROVED** [review]

**Pull request merged**